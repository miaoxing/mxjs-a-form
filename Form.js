import React from 'react';
import PropTypes from 'prop-types';
import ReactRouterPropTypes from 'react-router-prop-types';
import {withRouter} from 'react-router-dom';
import curUrl from '@mxjs/cur-url';
import {Form as AntdForm} from 'antd';
import $ from 'miaoxing';
import allTrim from 'all-trim';

/**
 * 在基础的表单上增加了
 *
 * 1. 简化 Formik 和 form 的层级结构
 * 2. 默认自动从后台读取数据
 * 3. 点击提交数据发送到后台
 * 4. 提交成功后跳转到相应页面
 */
class Form extends React.Component {
  static propTypes = {
    /**
     * 提交到后台的地址，默认自动识别为当前表单地址
     */
    url: PropTypes.string,

    /**
     * 提交到后台的方法
     */
    method: PropTypes.string,

    /**
     * 获取表单数据的后台地址
     */
    valuesUrl: PropTypes.oneOfType([PropTypes.string, PropTypes.bool]),

    /**
     * 提交成功后跳转的地址，默认为上一级页面
     */
    redirectUrl: PropTypes.string,

    /**
     * 加载数据后的回调
     */
    afterLoad: PropTypes.func,

    /**
     * 提交前回调
     */
    beforeSubmit: PropTypes.func,

    /**
     * 渲染子组件
     */
    render: PropTypes.func,

    /**
     * HTML form 元素的属性
     */
    formProps: PropTypes.object,

    /**
     * Formik 的属性
     */
    initialValues: PropTypes.object,

    formRef: PropTypes.oneOfType([PropTypes.func, PropTypes.shape({current: PropTypes.any})]),

    /**
     * 是否移除提交数据中的空白
     */
    trimSpaces: PropTypes.bool,

    children: PropTypes.oneOfType([PropTypes.node, PropTypes.func]),

    history: ReactRouterPropTypes.history.isRequired,
    location: ReactRouterPropTypes.location.isRequired,
    match: ReactRouterPropTypes.match.isRequired,
    staticContext: PropTypes.object,
  };

  static defaultProps = {
    formRef: React.createRef(),
    trimSpaces: true,
  };

  state = {
    initialValues: this.props.initialValues,
  };

  static getDerivedStateFromProps(props, state) {
    if (props.initialValues && props.initialValues !== state.initialValues) {
      return {
        initialValues: props.initialValues,
      };
    }
    return null;
  }

  handleSubmit = (values) => {
    if (this.props.beforeSubmit) {
      values = this.props.beforeSubmit(values);
    }

    if (this.props.trimSpaces) {
      values = allTrim(values);
    }

    $.http({
      ...this.getUrlAndMethod(),
      data: values,
      loading: true,
    }).then(ret => {
      $.ret(ret).suc(() => {
        this.redirect(this.getRedirectUrl());
      });
    });
  };

  redirect(url) {
    if (url !== this.props.history.location.pathname) {
      this.props.history.push(url);
    }
  }

  getUrlAndMethod() {
    const config = curUrl.apiFormUrlAndMethod();
    if (this.props.url) {
      config.url = this.props.url;
    }
    if (this.props.method) {
      config.method = this.props.method;
    }
    return config;
  }

  getRedirectUrl() {
    return this.props.redirectUrl || curUrl.index();
  }

  getValuesUrl() {
    if (typeof this.props.valuesUrl === 'undefined') {
      return curUrl.apiData();
    }

    return this.props.valuesUrl;
  }

  async componentDidMount() {
    if (this.props.initialValues) {
      return;
    }

    const valuesUrl = this.getValuesUrl();
    if (valuesUrl !== false) {
      // TODO 检查 ret
      const ret = await $.get(valuesUrl);
      if (this.props.afterLoad) {
        this.props.afterLoad(ret);
      }
      this.props.formRef.current.setFieldsValue(this.filterValues(ret.data));
    }
  }

  /**
   * 将输入项的值从 null 转换为空字符,因为 React input 值不允许为 null
   *
   * @param object data
   * @returns object
   */
  filterValues(data) {
    Object.keys(data).forEach(key => {
      if (data[key] === null) {
        data[key] = '';
      }
    });
    return data;
  }

  render() {
    const {
      children, formRef, valuesUrl, redirectUrl, beforeSubmit, staticContext, trimSpaces,
      history, location, match,
      ...props
    } = this.props;
    return (
      <AntdForm
        ref={formRef}
        initialValues={this.state.initialValues}
        labelCol={{span: 4}}
        wrapperCol={{span: 8}}
        onFinish={this.handleSubmit}
        {...props}
      >
        {children}
      </AntdForm>
    );
  }
}

export default withRouter(Form);

