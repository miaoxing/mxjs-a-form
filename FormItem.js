import { Component } from 'react';
import PropType from 'prop-types';
import {Form, InputNumber} from 'antd';
import Input from '@mxjs/a-input';

export default class extends Component {
  static propTypes = {
    children: PropType.node,
    type: PropType.string,
    name: PropType.oneOfType([PropType.string, PropType.array]),
    required: PropType.bool,
    messageVariables: PropType.objectOf(PropType.string),
  }

  renderChildren() {
    switch (true) {
      case !!this.props.children:
        return this.props.children;

      case this.props.type === 'plain':
        return ({getFieldValue}) => getFieldValue(this.props.name);

      case this.props.type === 'password':
        return <Input.Password/>;

      case this.props.type === 'hidden':
        return <Input type="hidden"/>;

      case this.props.type === 'number':
        return <InputNumber/>;

      case this.props.type === 'textarea':
        return <Input.TextArea/>;

      default:
        return <Input/>;
    }
  }

  render() {
    let {required, type, name, messageVariables = {}, ...props} = this.props;

    if (type === 'plain') {
      // 纯文本使用了自定义内容，需启用 shouldUpdate
      props.shouldUpdate = true;
      required = false;
    } else {
      // 纯文本不用 name（If you are using Form.Item as layout display, please remove `name` instead.）
      props.name = name;
    }

    if (!props.label) {
      messageVariables.label = '该项';
    }

    return <Form.Item
      noStyle={type === 'hidden'}
      required={required}
      rules={[{required: required}]}
      messageVariables={messageVariables}
      {...props}
    >
      {this.renderChildren()}
    </Form.Item>;
  }
}
