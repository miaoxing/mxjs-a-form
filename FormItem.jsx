import { Component } from 'react';
import PropType from 'prop-types';
import { Form, InputNumber } from 'antd';
import Input from '@mxjs/a-input';

export default class extends Component {
  static propTypes = {
    children: PropType.node,
    type: PropType.string,
    name: PropType.oneOfType([PropType.string, PropType.array]),
    required: PropType.bool,
    messageVariables: PropType.objectOf(PropType.string),
    controlProps: PropType.object,
  }

  static defaultProps = {
    type: 'text',
  }

  renderChildren() {
    switch (true) {
      case !!this.props.children:
        return this.props.children;

      case this.props.type === 'plain':
        return ({getFieldValue}) => getFieldValue(this.props.name);

      case this.props.type === 'password':
        return <Input.Password {...this.props.controlProps}/>;

      case this.props.type === 'hidden':
        return <Input type="hidden" {...this.props.controlProps}/>;

      case this.props.type === 'number':
        return <InputNumber {...this.props.controlProps}/>;

      case this.props.type === 'textarea':
        return <Input.TextArea {...this.props.controlProps}/>;

      default:
        return <Input type={this.props.type} {...this.props.controlProps}/>;
    }
  }

  render() {
    let {required, type, name, messageVariables = {}, controlProps, ...props} = this.props;

    if (type === 'plain') {
      // 纯文本使用了自定义内容，需启用 shouldUpdate
      props.shouldUpdate = true;
      required = false;
    } else {
      // 纯文本不用 name（If you are using Form.Item as layout display, please remove `name` instead.）
      props.name = name;
    }

    if (!props.label) {
      messageVariables.label = '该项';
    }

    return <Form.Item
      noStyle={type === 'hidden'}
      required={required}
      rules={[{required: required}]}
      messageVariables={messageVariables}
      {...props}
    >
      {this.renderChildren()}
    </Form.Item>;
  }
}
