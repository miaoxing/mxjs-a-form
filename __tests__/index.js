import 'core-js/stable';
import 'regenerator-runtime/runtime';

import React from 'react';
import Form from '../Form';
import FormItem from '../FormItem';
import {MemoryRouter, Router} from 'react-router';
import $ from 'miaoxing';
import {render, waitFor} from '@testing-library/react';
import {createPromise} from '@mxjs/test';
import {createMemoryHistory} from 'history';

describe('Form', () => {
  test('initialValues', async () => {
    const {container} = render(<MemoryRouter>
      <Form
        initialValues={{
          foo: 1,
          bar: 2,
        }}
      >
        <FormItem name="foo"/>
        <FormItem name="bar"/>
      </Form>
    </MemoryRouter>);

    expect(container.querySelector('#foo').value).toBe('1');
    expect(container.querySelector('#bar').value).toBe('2');
  });

  test('valuesUrl', async () => {
    const promise = createPromise();
    $.get = jest.fn().mockImplementationOnce(() => promise.resolve({
      data: {
        foo: 3,
        bar: 4,
      },
    }));

    const {container} = render(<MemoryRouter>
      <Form
        valuesUrl="test"
      >
        <FormItem name="foo"/>
        <FormItem name="bar"/>
      </Form>
    </MemoryRouter>);

    await promise;

    expect(container.querySelector('#foo').value).toBe('3');
    expect(container.querySelector('#bar').value).toBe('4');
  });

  test('beforeSubmit', async () => {
    const promise = createPromise();

    $.http = jest.fn().mockImplementationOnce(() => promise.resolve({
      code: 0,
      message: 'success',
    }));

    const form = React.createRef();
    render(<MemoryRouter>
      <Form
        url="test"
        initialValues={{
          foo: 1,
          bar: 2,
        }}
        formRef={form}
        beforeSubmit={(values) => {
          values.foo = 3;
          delete values.bar;
          return values;
        }}
      >
        <FormItem name="foo"/>
        <FormItem name="bar"/>
      </Form>
    </MemoryRouter>);

    form.current.submit();

    await promise;

    expect($.http).toHaveBeenCalledTimes(1);
    expect($.http).toMatchSnapshot();
  });

  test('beforeSubmit return false', async () => {
    const promise = createPromise();

    $.http = jest.fn().mockImplementationOnce(() => promise.resolve({
      code: 0,
      message: 'success',
    }));

    const form = React.createRef();
    render(<MemoryRouter>
      <Form
        url="test"
        initialValues={{
          foo: 1,
          bar: 2,
        }}
        formRef={form}
        beforeSubmit={() => {
          return false;
        }}
      >
        <FormItem name="foo"/>
        <FormItem name="bar"/>
      </Form>
    </MemoryRouter>);

    form.current.submit();

    expect($.http).toHaveBeenCalledTimes(0);
  });

  test('submit', async () => {
    const promise = createPromise();

    $.http = jest.fn().mockImplementationOnce(() => promise.resolve({
      code: 0,
      message: 'success',
    }));

    const form = React.createRef();
    render(<MemoryRouter>
      <Form
        url="test"
        initialValues={{
          foo: 1,
          bar: 2,
        }}
        formRef={form}
      >
        <FormItem name="foo"/>
        <FormItem name="bar"/>
      </Form>
    </MemoryRouter>);

    form.current.submit();

    await promise;

    expect($.http).toHaveBeenCalledTimes(1);
  });

  test('redirectUrl fn', async () => {
    const promise = createPromise();

    $.http = jest.fn().mockImplementationOnce(() => promise.resolve({
      code: 0,
      message: 'success',
    }));

    const history = createMemoryHistory();
    const form = React.createRef();
    render(<Router history={history}>
      <Form
        redirectUrl={(ret) => {
          return 'url/' + ret.message + '/' + ret.code;
        }}
        url="test"
        initialValues={{
          foo: 1,
          bar: 2,
        }}
        formRef={form}
      >
        <FormItem name="foo"/>
        <FormItem name="bar"/>
      </Form>
    </Router>);

    form.current.submit();

    await promise;

    await waitFor(() => {
      expect(history.location.pathname).toBe('/url/success/0');
    });

    expect($.http).toHaveBeenCalledTimes(1);
    expect($.http).toMatchSnapshot();
  });
});
