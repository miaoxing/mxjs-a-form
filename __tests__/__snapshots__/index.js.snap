// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form beforeSubmit 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "data": Object {
          "foo": 3,
        },
        "loading": true,
        "method": "POST",
        "url": "test",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Promise {
        "reject": [Function],
        "resolve": [Function],
      },
    },
  ],
}
`;

exports[`Form redirectUrl fn 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "data": Object {
          "bar": 2,
          "foo": 1,
        },
        "loading": true,
        "method": "POST",
        "url": "test",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Promise {
        "reject": [Function],
        "resolve": [Function],
      },
    },
  ],
}
`;
